### PR TITLE
Allowing pooling layers with different width and height

### PR DESCRIPTION
**Pull request by H4kor**
Solves issue #214

### CODE REVIEW

**Comment by nyanp** [comment]
@H4kor 
Great! I'll check your code :)

File: tiny_cnn/layers/average_pooling_layer.h [hunk 1]
```diff
@@ -95,7 +125,39 @@ class average_pooling_layer : public partial_connected_layer<Activation> {
...
+               pool_out_dim(in_width, pooling_size.first, stride.first) *
+               pool_out_dim(in_height, pooling_size.second, stride.second) * in_channels,
+               in_channels, in_channels, float_t(1) / (pooling_size.first * pooling_size.second)),
+          stride_width_(stride.first),
+          stride_height_(stride.second),
+          in_(in_width, in_height, in_channels),
+          out_(pool_out_dim(in_width, pooling_size, stride),
+               pool_out_dim(in_height, pooling_size, stride), in_channels),
+          w_(pooling_size, pooling_size, in_channels) {
+        if ((in_width % pooling_size.first)) {
```
Comment (nyanp) [inline tiny_cnn/layers/average_pooling_layer.h]: @H4kor 
We need to modify initializers of out_ and w_ to pass the compile.

```cpp
       out_(pool_out_dim(in_width, pooling_size.first, stride),
                pool_out_dim(in_height, pooling_size.second, stride), in_channels),
           w_(pooling_size.first, pooling_size.second, in_channels) {
```

File: tiny_cnn/layers/average_pooling_layer.h [hunk 2]
```diff
@@ -59,15 +59,44 @@ class average_pooling_layer : public partial_connected_layer<Activation> {
...
+    /**
+     * @param in_width     [in] width of input image
+     * @param in_height    [in] height of input image
+     * @param in_channels  [in] the number of input image channels(depth)
+     * @param pooling_size [in] factor by which to downscale
+     **/
+    average_pooling_layer(cnn_size_t in_width,
+                          cnn_size_t in_height,
+                          cnn_size_t in_channels,
+                          std::pair<cnn_size_t, cnn_size_t> pooling_size)
```
Comment (edgarriba) [inline tiny_cnn/layers/average_pooling_layer.h]: @H4kor thx for the contribution!

Could you switch `std::pair<cnn_size_t, cnn_size_t> pooling_size)` to two different variables?
For code consistency in our 'public' interfaces I think that will be better.

Comment (nyanp) [inline tiny_cnn/layers/average_pooling_layer.h]: +1 to the @edgarriba 's opinion :)

Comment (H4kor) [inline tiny_cnn/layers/average_pooling_layer.h]: @edgarriba if I use two `cnn_size_t` the ctor collides with the already existing ones.

Comment (edgarriba) [inline tiny_cnn/layers/average_pooling_layer.h]: I see. I think that proper solution could be having just two signatures: one with simple squared kernels and another with proper kernels definitions:
      
      average_pooling_layer(cnn_size_t in_width,
                            cnn_size_t in_height,
                            cnn_size_t in_channels,
                            cnn_size_t pooling_size,
                            cnn_size_t stride)

      average_pooling_layer(cnn_size_t in_width,
                            cnn_size_t in_height,
                            cnn_size_t in_channels,
                            cnn_size_t pooling_width,
                            cnn_size_t pooling_height,
                            cnn_size_t slide_width,
                            cnn_size_t slide_height)

/cc @nyanp @bhack

**Comment by edgarriba** [comment]
@nyanp Could you review this?

**Comment by nyanp** [comment]
It's nice PR, but I want the API not to use ```std::pair``` for consistency POV, as @edgarriba  said.

@H4kor 
Are you still interested in?

**Comment by jpauwels** [comment]
This appears to be implemented by #374 and could be closed imho

**Comment by Randl** [comment]
@edgarriba @nyanp Can you close this?

**Comment by bhack** [comment]
Ping

**Pull request closed without merging**